### PR TITLE
v3.5.0

### PR DESCRIPTION
**Pull request by ERussel**
- fix IRPromise bug when result is not propagated when immediate fulfilled
- fix IRPromise bug when result is not propagated when fulfilled in onError closure
- separate IRTransferAmount type added to add additional restrictions for amount in transactions (amount should be positive)

### CODE REVIEW

File: IrohaCommunication/Classes/Public/Promise/IRPromise.m [hunk 1]
```diff
@@ -44,42 +44,53 @@ + (nonnull instancetype)promiseWithResult:(nullable id)result {
...
-        self.errorHandler = block;
-        self.next = promise;
-
         if (self.isFulfilled) {
-            [self triggerResultProcessing];
+            if ([self.result isKindOfClass:[NSError class]]) {
+                promise = block(self.result);
+            }
+        } else {
+            promise = [[IRPromise alloc] init];
```
Comment (unklayman) [inline IrohaCommunication/Classes/Public/Promise/IRPromise.m]: Could be `[IRPromise new]`

File: IrohaCommunication/Classes/Public/Model/Common/IRAmount.m [hunk 2]
```diff
@@ -37,7 +34,7 @@ - (NSString *)description {
 
 @implementation IRAmountFactory
 
-+ (nullable id<IRAmount>)amountFromString:(nonnull NSString *)amount error:(NSError *_Nullable*_Nullable)error {
++ (nullable IRAmount*)concreteFromString:(nonnull NSString *)amount error:(NSError *_Nullable*_Nullable)error {
```
Comment (unklayman) [inline IrohaCommunication/Classes/Public/Model/Common/IRAmount.m]: \*

File: IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+static NSString * const DECIMAL_SEPARATOR = @".";
+
+@implementation NSDecimalNumber (IRAmount)
+
++ (nullable NSDecimalNumber*)decimalNumberWithAmountValue:(nonnull NSString*)value {
+    return [NSDecimalNumber decimalNumberWithString:value
+                                             locale:@{NSLocaleDecimalSeparator: DECIMAL_SEPARATOR}];
+}
+
++ (nullable NSDecimalNumber*)decimalNumberWithAmount:(id<IRAmount>)amount {
```
Comment (unklayman) [inline IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m]: \*

File: IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+* SPDX-License-Identifier: Apache-2.0
+*/
+
+#import "NSDecimalNumber+IRAmount.h"
+
+static NSString * const DECIMAL_SEPARATOR = @".";
+
+@implementation NSDecimalNumber (IRAmount)
+
++ (nullable NSDecimalNumber*)decimalNumberWithAmountValue:(nonnull NSString*)value {
```
Comment (unklayman) [inline IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m]: \*

File: IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#import <Foundation/Foundation.h>
+#import "IRAmount.h"
+
+@interface NSDecimalNumber (IRAmount)
+
++ (nullable NSDecimalNumber*)decimalNumberWithAmountValue:(nonnull NSString*)value;
++ (nullable NSDecimalNumber*)decimalNumberWithAmount:(nonnull id<IRAmount>)amount;
```
Comment (unklayman) [inline IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.h]: \*

File: IrohaCommunication/Classes/Public/Promise/IRPromise.m (hunk 1 above)
Comment (ERussel) [reply IrohaCommunication/Classes/Public/Promise/IRPromise.m]: We need to refactor all the code though. For now let's leave current style.

File: IrohaCommunication/Classes/Public/Model/Common/IRAmount.m (hunk 2 above)
Comment (ERussel) [reply IrohaCommunication/Classes/Public/Model/Common/IRAmount.m]: fixed

File: IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m (hunk 3 above)
Comment (ERussel) [reply IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m]: fixed

File: IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m (hunk 4 above)
Comment (ERussel) [reply IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.m]: fixed

File: IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.h (hunk 5 above)
Comment (ERussel) [reply IrohaCommunication/Classes/Public/Extensions/NSDecimalNumber+IRAmount.h]: fixed

**Pull request merged**